### ISSUE TITLE
Raise exception if `request.tm` is accessed for request opted out by `activate_hook`

### ISSUE THREAD

**Issue by nickstenning** [Pylons/pyramid_tm#45]
If you use an `activate_hook` to opt certain requests out of the transaction manager tween, it is currently still possible to get a reference to the request transaction manager using `request.tm`. 

This isn't usually a big problem, but if instead of using a threadlocal transaction manager, you are using a per-request tm, say

```python
config.add_settings({
   "tm.manager_hook": lambda request: transaction.TransactionManager(),
})
```

and then registering the session with the transaction manager using `zope.sqlalchemy.register(session, transaction_manager=request.tm)`, then it's possible to end up in a situation where the session is registered with the transaction manager but the transaction manager isn't hooked up to the request lifecycle.

This wouldn't be the end of the world, but it results in references to `id(session)` not being cleaned up from [the `_SESSION_STATE` threadlocal](https://github.com/zopefoundation/zope.sqlalchemy/blob/9b70ff1cdbebb328ed287183112b696c1184ec61/src/zope/sqlalchemy/datamanager.py#L48) in `zope.sqlalchemy`.

In an environment where a new `Session` instance is created for each request, this results in a (small) memory leak. Much more importantly, it is very easy to get into a situation where a brand-new session is created with an `id(session)` that's already in `_SESSION_STATE`, resulting in `zope.sqlalchemy.register` silently failing to register the session with the transaction manager.

Of course, part of the problem here is that of usage: why would you call `zope.sqlalchemy.register(...)` in a request that you opted out of the transaction manager machinery. But I'd argue that this is a bit of a footgun, and `pyramid_tm` could prevent this class of bug entirely if the `tm` request property raised an `AttributeError` (or a custom exception type) when used in the context of a request for which `activate_hook` returned False.

I'm happy to put together a PR to implement this behaviour, but thought I should open this issue for feedback on the idea before I did so. Thoughts?

**Comment by mmerickel** [comment]
I'd accept a PR that raised an `AttributeError` for `request.tm` if the TM is inactive. To avoid calling the hook more than once it might also be helpful to put something like `tm.active` in the environ (and remove it on egress) that can be checked. This would be similar to https://github.com/Pylons/pyramid_tm/blob/92d746081d65926082359c3db923928546d16b69/pyramid_tm/__init__.py#L45

**Comment by mmerickel** [comment]
Closed via #46.
